Sources/_StringProcessing/RegexDSL/AnyRegexOutput.swift

```diff
@@ -0,0 +1,128 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2021-2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+import _MatchingEngine
+
+extension Regex where Output == AnyRegexOutput {
+  public init(_ pattern: String) throws {
+    self.init(ast: try parse(pattern, .traditional))
+  }
+}
+
+extension Regex.Match where Output == AnyRegexOutput {
+  // Ensures `.0` always refers to the whole match.
+  public subscript(
+    dynamicMember keyPath: KeyPath<(Substring, _doNotUse: ()), Substring>
+  ) -> Substring {
+    input[range]
+  }
+}
+
+public struct AnyRegexOutput {
+  let input: String
+  fileprivate let _elements: [ElementRepresentation]
+
+  /// The underlying representation of the element of a type-erased regex
+  /// output.
+  fileprivate struct ElementRepresentation {
+    /// The depth of `Optioals`s wrapping the underlying value. For example,
+    /// `Substring` has optional depth `0`, and `Int??` has optional depth `2`.
+    let optionalDepth: Int
```

Comment:
Unrelated to this PR, did we decide on an optional nesting story?

Reply:
There’s no non-factorial solution to optional flattening this year, unfortunately. So we’ll have nested optionals.